device_backends/src/InterruptControllerHandler.cc

```diff
@@ -0,0 +1,130 @@
+      }
+      idAsString.pop_back(); // remove last ":"
+      throw ChimeraTK::logic_error("Unknown interrupt controller ID " + idAsString);
+    }
+    auto creatorFunctionIter = _creatorFunctions.find(name);
+    if(creatorFunctionIter == _creatorFunctions.end()) {
+      throw ChimeraTK::logic_error("Unknown interrupt controller type \"" + name + "\"");
+    }
+    return creatorFunctionIter->second(this, controllerID, description, std::move(parent));
+  }
+
+  //*********************************************************************************************************************/
+  boost::shared_ptr<DeviceBackend> InterruptControllerHandlerFactory::getBackend() {
+    return boost::dynamic_pointer_cast<DeviceBackend>(_backend->shared_from_this());
+  }
+
+  //*********************************************************************************************************************/
+  InterruptControllerHandler::InterruptControllerHandler(InterruptControllerHandlerFactory* controllerHandlerFactory,
+      std::vector<uint32_t> controllerID, boost::shared_ptr<TriggerDistributor> parent)
+  : _backend(controllerHandlerFactory->getBackend()), _controllerHandlerFactory(controllerHandlerFactory),
+    _id(std::move(controllerID)), _parent(std::move(parent)), _asyncDomain(_parent->getAsyncDomain()) {}
+
+  //*********************************************************************************************************************/
+  template<typename DistributorType>
+  boost::shared_ptr<DistributorType> InterruptControllerHandler::getDistributorRecursive(
+      std::vector<uint32_t> const& interruptID) {
+    // assert(false); // FIXME: needs container lock!
+    assert(!interruptID.empty());
+    auto qualifiedInterruptId = _id;
+    qualifiedInterruptId.push_back(interruptID.front());
+
+    // we can't use try_emplace because the map contains weak pointers
+    boost::shared_ptr<TriggerDistributor> distributor;
+    auto distributorIter = _distributors.find(interruptID.front());
+    if(distributorIter == _distributors.end()) {
+      distributor = boost::make_shared<TriggerDistributor>(
+          _backend, _controllerHandlerFactory, qualifiedInterruptId, shared_from_this(), _asyncDomain);
+      _distributors[interruptID.front()] = distributor;
+      if(_asyncDomain->_isActive) {
+        // Creating a new version here is correct. Nothing has been distributed to any accessor connected to this
+        // sub-interrupt yet because we just created this distributor.
+        distributor->activate(nullptr, {});
+      }
+    }
+    else {
+      distributor = distributorIter->second.lock();
+      if(!distributor) {
+        distributor = boost::make_shared<TriggerDistributor>(
+            _backend, _controllerHandlerFactory, qualifiedInterruptId, shared_from_this(), _asyncDomain);
+        distributorIter->second = distributor;
+        if(_asyncDomain->_isActive) {
+          distributor->activate(nullptr, {});
+        }
+      }
+    }
```

Comment:
```suggestion
    if(distributorIter != _distributors.end()) {
      distributor = distributorIter->second.lock();
    }
    if(!distributor) {
      distributor = boost::make_shared<TriggerDistributor>(
          _backend, _controllerHandlerFactory, qualifiedInterruptId, shared_from_this(), _asyncDomain);
      _distributors[interruptID.front()] = distributor;
      if(_asyncDomain->_isActive) {
        // Creating a new version here is correct. Nothing has been distributed to any accessor connected to this
        // sub-interrupt yet because we just created this distributor.
        distributor->activate(nullptr, {});
      }
    }
```

Yes, this is slightly less efficient in case the iterator points to a distributor which went away, since the square bracket operator needs to lookup the map again, but this is negligible.

Reply:
I never liked the code duplication anyway. I just did not come up with a better solution. Thanks!